src/lpython/semantics/python_comptime_eval.h

```diff
@@ -163,6 +167,66 @@ struct PythonIntrinsicProcedures {
                 ASRUtils::type_to_str_python(arg_type) + "'", loc);
         }
     }
+    static ASR::expr_t *eval_upper(Allocator &al, const Location &loc, Vec<ASR::expr_t*> &args) {
+        LFORTRAN_ASSERT(ASRUtils::all_args_evaluated(args));
+        if (args.size() != 1) {
+            throw SemanticError("upper() takes exactly one argument (" +
+                std::to_string(args.size()) + " given)", loc);
+        }
+        ASR::ttype_t* str_type = ASRUtils::expr_type(args[0]);
+        ASR::expr_t *arg = args[0];
+        ASR::ttype_t* arg_type = ASRUtils::expr_type(arg);
+        if (ASRUtils::is_character(*arg_type)) {
+            char* c = ASR::down_cast<ASR::StringConstant_t>(arg)->m_s;
+            std::string s = (int (*)(int))std::toupper(std::string(c));
+            return ASR::down_cast<ASR::expr_t>(ASR::make_StringConstant_t(al, loc, s2c(al, s), str_type));
+        }
+        else {
+            throw SemanticError("upper() only works on strings", loc);
+        }
+    }
+
+    static ASR::expr_t *eval_lower(Allocator &al, const Location &loc, Vec<ASR::expr_t*> &args) {
+        LFORTRAN_ASSERT(ASRUtils::all_args_evaluated(args));
+        if (args.size() != 1) {
+            throw SemanticError("lower() takes exactly one argument (" +
+                std::to_string(args.size()) + " given)", loc);
+        }
+        ASR::ttype_t* str_type = ASRUtils::expr_type(args[0]);
+        ASR::expr_t *arg = args[0];
+        ASR::ttype_t* arg_type = ASRUtils::expr_type(arg);
+
+        if (ASRUtils::is_character(*arg_type)) {
+            char* c = ASR::down_cast<ASR::StringConstant_t>(arg)->m_s;
+            std::string s = (int (*)(int))std::tolower(std::string(c));
+            return ASR::down_cast<ASR::expr_t>(ASR::make_StringConstant_t(al, loc, s2c(al, s), str_type));
+        }
+        else {
+            throw SemanticError("lower() only works on strings", loc);
+        }
+    }
+    static ASR::expr_t *eval_capitalize(Allocator &al, const Location &loc, Vec<ASR::expr_t*> &args) {
+        LFORTRAN_ASSERT(ASRUtils::all_args_evaluated(args));
+        if (args.size() != 1) {
+            throw SemanticError("capitalize() takes exactly one argument (" +
+                std::to_string(args.size()) + " given)", loc);
+        }
+        ASR::ttype_t* str_type = ASRUtils::expr_type(args[0]);
+        ASR::expr_t *arg = args[0];
+        ASR::ttype_t* arg_type = ASRUtils::expr_type(arg);
+
+        if (ASRUtils::is_character(*arg_type)) {
+            char* c = ASR::down_cast<ASR::StringConstant_t>(arg)->m_s;
+            std::string s = std::string(c);
+            s[0] = ::toupper(s[0]);
+            std::transform(s.begin()+1, s.end(), s.begin()+1, (int (*)(int))std::tolower);
```

Comment:
Let's not do `(int (*)(int))`. If you need a C string, let's write some wrapper callback.